File: leetcode/word_break.py
# Given a string s and a dictionary of words dict, determine if s can be 
# segmented into a space-separated sequence of one or more dictionary words.
# Time Complexity: O(n^2 + m) in average, m is the size of dictionary
# Space Complexity: O(n + m)

class Solution:
    # @param s, a string
    # @param wordDict, a set<string>
    # @return a boolean
    def wordBreak(self, s, wordDict):
        trie = {}
        for word in wordDict:
            node = trie
            for c in word[::-1]:
                node = node.setdefault(c, {})
            node[None] = None
        F = [False] * (len(s) + 1)
        F[0] = True
        for i in xrange(len(s)):
            j, node = i, trie
            while not F[i + 1] or j >= 0:
                if s[j] not in node: break
                node = node[s[j]]
                if F[j] and None in node: F[i + 1] = True
                j -= 1
        return F[len(s)]
                

if __name__ == "__main__":
    solution = Solution()
    print solution.wordBreak("leetcode", ["leet", "code"])

File: leetcode/word_break.js
/* Given a string s and a dictionary of words dict, determine if s can be 
 * segmented into a space-separated sequence of one or more dictionary words.
 * Time Complexity: O(n^2 + m) in average, m is the size of dictionary
 * Space Complexity: O(n + m)
 */

/**
 * @param {string} s
 * @param {set<string>} wordDict
 *   Note: wordDict is a Set object, see:
 *   https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set
 * @return {boolean}
 */
var wordBreak = function(s, wordDict) {
    var trie = new Map();
    for (var word of wordDict) {
        var node = trie;
        for (var i = word.length - 1; i >= 0; --i) {
            var c = word.charAt(i);
            if (!node.has(c))
                node.set(c, new Map());
            node = node.get(c);
        }
        node.set(null, null);
    }
    var F = new Array(s.length + 1);
    F.fill(false);
    F[0] = true;
    for (i = 0; i < s.length; ++i) {
        node = trie;
        for (var j = i; !F[i + 1] || j >= 0; --j) {
            if (!node.has(s.charAt(j)))
                break;
            node = node.get(s.charAt(j));
            if (F[j] && node.has(null))
                F[i + 1] = true;
        }
    }
    return F[s.length];
};
